synapse/storage/databases/state/bg_updates.py

```diff
@@ -173,25 +210,83 @@ def _get_state_groups_from_groups_txn(
+                # We default to `[-1]` just to fill in the query with something that
+                # will have no effect but not bork our query when it would be empty
+                # otherwise
+                -1
+            }
+            for group in sorted_groups:
                 args: List[Union[int, str]] = [group]
+                args.extend(state_groups_we_have_already_fetched)
                 args.extend(overall_select_query_args)
 
-                txn.execute(sql % (overall_select_clause,), args)
+                state_groups_we_have_already_fetched_string = ", ".join(
+                    ["?::bigint"] * len(state_groups_we_have_already_fetched)
+                )
+
+                txn.execute(
+                    sql
+                    % (
+                        state_groups_we_have_already_fetched_string,
+                        overall_select_clause % (main_select_clause,),
+                    ),
+                    args,
+                )
+
+                # The first row is always our special `state_group_reached` row which
+                # tells us if we linked up to any other existing state_group that we
+                # already fetched and if so, which one we linked up to (see the `UNION
+                # ALL` above which drives this special row)
+                first_row = txn.fetchone()
+                if first_row:
+                    _, _, _, state_group_reached = first_row
+
+                partial_state_map_for_state_group: MutableStateMap[str] = {}
                 for row in txn:
-                    typ, state_key, event_id = row
+                    typ, state_key, event_id, _state_group = row
                     key = (intern_string(typ), intern_string(state_key))
-                    results[group][key] = event_id
+                    partial_state_map_for_state_group[key] = event_id
+
+                # If we see a state_group edge link to a previous state_group that we
+                # already fetched from the database, link up the base state to the
+                # partial state we retrieved from the database to build on top of.
+                if state_group_reached in results:
+                    resultant_state_map = dict(results[state_group_reached])
+                    resultant_state_map.update(partial_state_map_for_state_group)
+
+                    results[group] = resultant_state_map
+                else:
+                    # It's also completely normal for us not to have a previous
+                    # state_group to build on top of if this is the first group being
+                    # processed or we are processing a bunch of groups from different
+                    # rooms which of course will never link together (competely
+                    # different DAGs).
+                    results[group] = partial_state_map_for_state_group
+
+                state_groups_we_have_already_fetched.add(group)
+
```

Comment:
> In the sample request above, simulating how the app would make the queries with the changes in this PR, we don't actually see that much benefit because it turns out that there isn't much `state_group` sharing amongst events. That 14s SQL query is just fetching the state at each of the `prev_events` of one event and only one seems to share `state_groups`.
> 
> [...]
> 
> Initially, I thought the lack of sharing was quite strange but this is because of the state_group snapshotting feature where if an event requires more than 100 hops on `state_event_edges`, then a Synaspe will create a new `state_group` with a snapshot of all of that state. It seems like this isn't done very efficiently though. [Relevant docs](https://github.com/matrix-org/synapse/blob/develop/docs/development/room-dag-concepts.md#state-groups).
> 
> And it turns out the event is an `org.matrix.dummy_event` which Synapse automatically puts in the DAG to resolve outstanding forward extremities and [these events aren't even shown to clients](https://github.com/matrix-org/synapse/blob/ad50510a06d035a674f0eeed5db5dd3060bc0b1c/synapse/visibility.py#L494-L495) so we don't even need to waste time waiting for them to backfill. **Tracked by** https://github.com/matrix-org/synapse/issues/15632
> 
> Generally, I think this PR could bring great gains in conjunction to running [some sort of state compressor](https://github.com/matrix-org/rust-synapse-compress-state) over the database to get a lot more sharing. In addition to trying to fix the online state_group snapshotting logic to be smarter. I don't know how the existing state_compressors work but I imagine we could create snapshots and bucket for years -> months -> weeks -> days -> hours -> individual events and create new `state_group` chains which utilize these from biggest to smallest to get maximal sharing.
>
> *-- "After" section of the PR description*

This hasn't made as big of an impact as I thought it would for that type of request. Are we still interested in a change like this?

It seems like our `state_group` sharing is realllly sub-par and the way that `state_groups` can only have a max of 100 hops puts an upper limit on how much gain this PR can give. I didn't anticipate that's how `state_groups` worked and thought it was one `state_group` per-state-change which it is until it starts doing snapshots.

Maybe it's more interesting to improve our `state_group` logic to be much smarter first and we could re-visit something like this. Or look into the state compressor stuff to optimize our backlog which would help for the Matrix Public Archive.